### PR TITLE
introduce hgraph

### PR DESCRIPTION
**Pull request by LHT129**
- add basic implement for hgraph(hnswsq version)
- add one test for it

### CODE REVIEW

File: examples/cpp/example_hgraph.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            ->Float32Vectors(data.get())
+            ->Owner(false);
+        if (const auto num = index.value()->Build(dataset); num.has_value()) {
+            std::cout << "After Build(), Index constains: " << index.value()->GetNumElements()
+                      << std::endl;
+        } else if (num.error().type == vsag::ErrorType::INTERNAL_ERROR) {
+            std::cerr << "Failed to build index: internalError" << std::endl;
+            exit(-1);
+        }
+    }
```
Comment (inabao) [inline examples/cpp/example_hgraph.cpp]: add the search process and verify the result

File: examples/cpp/CMakeLists.txt [hunk 2]
```diff
@@ -21,3 +21,6 @@ target_link_libraries(custom_logger vsag)
 
 add_executable (vsag_ext_example vsag_ext_example.cpp)
 target_link_libraries(vsag_ext_example vsag)
+
+add_executable (example_hgraph example_hgraph.cpp)
+target_link_libraries(example_hgraph vsag -fopenmp gomp)
```
Comment (inabao) [inline examples/cpp/CMakeLists.txt]: why do you link the openmp here？

File: examples/cpp/example_hgraph.cpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+// limitations under the License.
+
+#include <cstdint>
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <nlohmann/json.hpp>
+#include <random>
+
+#include "local_file_reader.h"
```
Comment (inabao) [inline examples/cpp/example_hgraph.cpp]: why do you include the file here?

File: src/factory/factory.cpp [hunk 4]
```diff
@@ -81,6 +83,30 @@ Factory::CreateIndex(const std::string& origin_name,
                                              params.use_opq,
                                              params.use_bsa,
                                              params.use_async_io);
+        } else if (name == INDEX_HGRAPH) {
+            auto json_param = nlohmann::json::parse(parameters);
+            IndexCommonParam param;
+            CHECK_ARGUMENT(json_param.contains(PARAMETER_METRIC_TYPE),
```
Comment (inabao) [inline src/factory/factory.cpp]: use CHECK_ARGUMENT in parsing process rather than here.

File: src/factory/factory.cpp [hunk 5]
```diff
@@ -81,6 +83,30 @@ Factory::CreateIndex(const std::string& origin_name,
...
+                param.metric_ = vsag::MetricType::METRIC_TYPE_L2SQR;
+            } else if (json_param[PARAMETER_METRIC_TYPE] == METRIC_IP) {
+                param.metric_ = vsag::MetricType::METRIC_TYPE_IP;
+            } else if (json_param[PARAMETER_METRIC_TYPE] == METRIC_COSINE) {
+                param.metric_ = vsag::MetricType::METRIC_TYPE_COSINE;
+            }
+            if (allocator != nullptr) {
+                param.allocator_ = allocator;
+            } else {
+                param.allocator_ = new DefaultAllocator();
```
Comment (inabao) [inline src/factory/factory.cpp]: replace it with DefaultAllocator::Instance();

File: src/index/hgraph_index.cpp [hunk 6]
```diff
@@ -0,0 +1,676 @@
...
+        candidate_set.pop();
+
+        auto current_node_id = current_node_pair.second;
+        {
+            std::shared_lock<std::shared_mutex> lock(neighbors_mutex_[current_node_id]);
+            graph->GetNeighbors(current_node_id, neighbors);
+        }
+        if (!neighbors.empty()) {
+            flatten->Prefetch(neighbors[0]);
+#ifdef USE_SSE
```
Comment (inabao) [inline src/index/hgraph_index.cpp:345]: what's the difference between the `flatten->Prefetch` and the following prefetch process.

File: tests/performance/test_performance.cpp [hunk 7]
```diff
@@ -58,6 +58,48 @@ main(int argc, char* argv[]) {
     std::string build_parameters = argv[4];
     std::string search_parameters = argv[5];
 
+    if (build_parameters == "hello hgraph") {
```
Comment (inabao) [inline tests/performance/test_performance.cpp]: what's the mean of `hello hgraph`

Comment (LHT129) [reply tests/performance/test_performance.cpp]: the origin param is too long，so I mark another name to do this, I will reset the changes in test_performance and examples

File: examples/cpp/example_hgraph.cpp (hunk 1 above)
Comment (LHT129) [reply examples/cpp/example_hgraph.cpp]: delete example. I will add new example use new params

File: examples/cpp/CMakeLists.txt (hunk 2 above)
Comment (LHT129) [reply examples/cpp/CMakeLists.txt]: ditto

File: examples/cpp/example_hgraph.cpp (hunk 3 above)
Comment (LHT129) [reply examples/cpp/example_hgraph.cpp]: ditto

File: src/factory/factory.cpp (hunk 4 above)
Comment (LHT129) [reply src/factory/factory.cpp]: done

File: src/factory/factory.cpp (hunk 5 above)
Comment (LHT129) [reply src/factory/factory.cpp]: done

File: src/index/hgraph_index.cpp (hunk 6 above)
Comment (LHT129) [reply src/index/hgraph_index.cpp:345]: this prefetch vec data, the following prefetch visit_list

**Review by inabao — APPROVED** [review]
LGTM

File: src/factory/factory.cpp [hunk 8]
```diff
@@ -81,6 +83,19 @@ Factory::CreateIndex(const std::string& origin_name,
                                              params.use_opq,
                                              params.use_bsa,
                                              params.use_async_io);
+        } else if (name == INDEX_HGRAPH) {
+            auto json_param = nlohmann::json::parse(parameters);
```
Comment (jiaweizone) [inline src/factory/factory.cpp]: json_param -> params ?

File: src/factory/factory.cpp [hunk 9]
```diff
@@ -81,6 +83,19 @@ Factory::CreateIndex(const std::string& origin_name,
...
+            auto common_param = IndexCommonParam::CheckAndCreate(json_param);
+
+            if (allocator != nullptr) {
+                common_param.allocator_ = allocator;
+            } else {
+                common_param.allocator_ = DefaultAllocator::Instance().get();
+            }
+            logger::debug("created a hgraph index");
+            auto result = std::make_shared<HGraphIndex>(json_param["index_param"], common_param);
```
Comment (jiaweizone) [inline src/factory/factory.cpp]: minor: result -> index_obj ?

File: src/index/hgraph_index.cpp [hunk 10]
```diff
@@ -0,0 +1,675 @@
...
+#include "data_cell/sparse_graph_datacell.h"
+#include "hnsw.h"
+#include "hnsw_zparameters.h"
+
+namespace vsag {
+BinarySet
+empty_binaryset() {
+    const std::string empty_str = "EMPTY_INDEX";
+    size_t num_bytes = empty_str.length();
+    std::shared_ptr<int8_t[]> bin(new int8_t[num_bytes]);
```
Comment (jiaweizone) [inline src/index/hgraph_index.cpp:30]: can define an static global `s_empty_binaryset' object ?

File: src/index/hgraph_index.cpp [hunk 11]
```diff
@@ -0,0 +1,675 @@
...
+        .data = bin,
+        .size = num_bytes,
+    };
+    BinarySet bs;
+    bs.Set(BLANK_INDEX, b);
+
+    return bs;
+}
+
+HGraphIndex::HGraphIndex(const nlohmann::json& json_obj,
```
Comment (jiaweizone) [inline src/index/hgraph_index.cpp]: json_obj -> index_params ?

Comment (jiaweizone) [inline src/index/hgraph_index.cpp]: nlohmann::json -> JsonType ?

File: src/index/hgraph_index.cpp [hunk 12]
```diff
@@ -0,0 +1,675 @@
...
+    this->pool_ = std::make_shared<hnswlib::VisitedListPool>(
+        1, this->bottom_graph_->MaxCapacity(), allocator_);
+
+    if (this->build_thread_count_ > 1) {
+        this->build_pool_ = std::make_unique<progschj::ThreadPool>(this->build_thread_count_);
+    }
+}
+
+tl::expected<std::vector<int64_t>, Error>
+HGraphIndex::build(const DatasetPtr& base) {
```
Comment (jiaweizone) [inline src/index/hgraph_index.cpp]: base -> data ?

File: src/index/hgraph_index.cpp [hunk 13]
```diff
@@ -0,0 +1,675 @@
...
+    return this->add(base);
+}
+
+tl::expected<std::vector<int64_t>, Error>
+HGraphIndex::add(const DatasetPtr& base) {
```
Comment (jiaweizone) [inline src/index/hgraph_index.cpp]: ditto

File: src/index/hgraph_index.cpp [hunk 14]
```diff
@@ -0,0 +1,675 @@
...
+        this->basic_flatten_codes_->Train(base->GetFloat32Vectors(), base->GetNumElements());
+        this->basic_flatten_codes_->BatchInsertVector(base->GetFloat32Vectors(),
+                                                      base->GetNumElements());
+        if (use_reorder_) {
+            this->high_precise_codes_->Train(base->GetFloat32Vectors(), base->GetNumElements());
+            this->high_precise_codes_->BatchInsertVector(base->GetFloat32Vectors(),
+                                                         base->GetNumElements());
+        }
+        this->hnsw_add(base);
+        return failed_ids;
```
Comment (jiaweizone) [inline src/index/hgraph_index.cpp]: `failed_ids` not touch ?

**Review by jiaweizone — COMMENTED** [review]
PART-I

File: src/index/hgraph_index.cpp (hunk 10 above)
Comment (LHT129) [reply src/index/hgraph_index.cpp:30]: this is the same as hnsw, how to process empty index may need discuss

File: src/factory/factory.cpp (hunk 8 above)
Comment (LHT129) [reply src/factory/factory.cpp]: done

File: src/factory/factory.cpp (hunk 9 above)
Comment (LHT129) [reply src/factory/factory.cpp]: done

File: src/index/hgraph_index.cpp (hunk 11 above)
Comment (LHT129) [reply src/index/hgraph_index.cpp]: done

Comment (LHT129) [reply src/index/hgraph_index.cpp]: rename in another pr

File: src/index/hgraph_index.cpp (hunk 13 above)
Comment (LHT129) [reply src/index/hgraph_index.cpp]: done

File: src/index/hgraph_index_test.cpp [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+
+    index = std::make_shared<vsag::HGraphIndex>(json_param["index_param"], param);
+    index->Init();
+    std::ifstream infile(filename.c_str(), std::ios::binary);
+    IOStreamReader reader(infile);
+    index->deserialize(reader);
+    func();
+
+    infile.close();
+}
```
Comment (wxyucs) [inline src/index/hgraph_index_test.cpp]: add a blank line at the end of file

File: src/index/hgraph_index_test.cpp [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+#include <catch2/catch_test_macros.hpp>
+#include <fstream>
+#include <memory>
+#include <nlohmann/json.hpp>
+#include <vector>
+
+#include "fixtures.h"
+#include "vsag/errors.h"
+
+TEST_CASE("build with allocator [ut][hgraphindex]") {
```
Comment (wxyucs) [inline src/index/hgraph_index_test.cpp]: incorrect TEST_CASE format, use TEST_CASE("test_name", "[tag1][tag2]")

**Review by wxyucs — CHANGES_REQUESTED** [review]

File: src/index/hgraph_index.cpp (hunk 14 above)
Comment (LHT129) [reply src/index/hgraph_index.cpp]: add todo to fulfill the failed_ids

File: src/index/hgraph_index_test.cpp (hunk 16 above)
Comment (LHT129) [reply src/index/hgraph_index_test.cpp]: done

File: src/index/hgraph_index_test.cpp (hunk 15 above)
Comment (LHT129) [reply src/index/hgraph_index_test.cpp]: done

File: src/index/hgraph_index.h [hunk 17]
```diff
@@ -0,0 +1,282 @@
...
+        constexpr bool
+        operator()(std::pair<float, InnerIdType> const& a,
+                   std::pair<float, InnerIdType> const& b) const noexcept {
+            return a.first < b.first;
+        }
+    };
+
+    using MaxHeap = std::priority_queue<std::pair<float, InnerIdType>,
+                                        Vector<std::pair<float, InnerIdType>>,
+                                        CompareByFirst>;
```
Comment (wxyucs) [inline src/index/hgraph_index.h:46]: move these structure definitions to the private section if they will not be used from otherwhere

File: src/index/hgraph_index.h [hunk 18]
```diff
@@ -0,0 +1,282 @@
...
+    SetBuildThreadsCount(uint64_t count) {
+        this->build_thread_count_ = count;
+        this->build_pool_->set_pool_size(count);
+    }
+
+    void
+    serialize(StreamWriter& writer) const;
+
+    void
+    deserialize(StreamReader& reader);
```
Comment (wxyucs) [inline src/index/hgraph_index.h:167]: what is the reason to make these two functions public

File: src/index/hgraph_index.h [hunk 19]
```diff
@@ -0,0 +1,282 @@
...
+private:
+    std::default_random_engine level_generator_{2021};
+    double mult_{1.0};
+
+    Allocator* allocator_{nullptr};
+
+    UnorderedMap<LabelType, InnerIdType> label_lookup_;
+    mutable std::shared_mutex label_lookup_mutex_{};  // lock for label_lookup_
+
+    InnerIdType enter_point_id_{std::numeric_limits<InnerIdType>::max()};
```
Comment (wxyucs) [inline src/index/hgraph_index.h]: entrypoint?

File: src/index/hgraph_index.cpp [hunk 20]
```diff
@@ -0,0 +1,675 @@
...
+            auto end = std::min(j * task_size + task_size, total);
+            this->build_pool_->enqueue(build_func, j * task_size, end);
+        }
+        this->build_pool_->wait_until_nothing_in_flight();
+    } else {
+        build_func(0, total);
+    }
+}
+
+std::shared_ptr<GraphInterface>
```
Comment (wxyucs) [inline src/index/hgraph_index.cpp]: GraphInterfacePtr

File: src/index/hgraph_index.cpp [hunk 21]
```diff
@@ -0,0 +1,675 @@
...
+                                                 bottom_graph_->MaximumDegree() / 2);
+}
+
+HGraphIndex::MaxHeap
+HGraphIndex::search_one_graph(const float* query,
+                              const std::shared_ptr<GraphInterface>& graph,
+                              const std::shared_ptr<FlattenInterface>& flatten,
+                              InnerIdType ep,
+                              uint64_t ef,
+                              BaseFilterFunctor* isIdAllowed) const {
```
Comment (wxyucs) [inline src/index/hgraph_index.cpp]: is_id_allowed

File: src/index/hgraph_index.h [hunk 22]
```diff
@@ -0,0 +1,282 @@
...
+    deserialize(const ReaderSet& reader_set);
+
+    tl::expected<void, Error>
+    deserialize(const BinarySet& binary_set);
+
+    tl::expected<void, Error>
+    deserialize(std::istream& in_stream);
+
+    tl::expected<float, Error>
+    calcDistanceById(const float* vector, int64_t id) const;
```
Comment (wxyucs) [inline src/index/hgraph_index.h]: calc_distance_by_id

File: src/index/hgraph_index.cpp [hunk 23]
```diff
@@ -0,0 +1,675 @@
...
+                                          std::shared_ptr<GraphInterface> graph,
+                                          std::shared_ptr<FlattenInterface> flatten,
+                                          bool is_update) {
+    const size_t max_size = graph->MaximumDegree();
+    this->select_edges_by_heuristic(top_candidates, max_size, flatten);
+    if (top_candidates.size() > max_size)
+        throw std::runtime_error(
+            "Should be not be more than max_size candidates returned by the heuristic");
+
+    Vector<InnerIdType> selectedNeighbors(allocator_);
```
Comment (wxyucs) [inline src/index/hgraph_index.cpp]: selected_neighbors

File: src/index/hgraph_index.h (hunk 18 above)
Comment (LHT129) [reply src/index/hgraph_index.h:167]: currently for unittest, I will make it private, after the whole test finished

File: src/index/hgraph_index.h (hunk 22 above)
Comment (LHT129) [reply src/index/hgraph_index.h]: done

File: src/index/hgraph_index.cpp (hunk 21 above)
Comment (LHT129) [reply src/index/hgraph_index.cpp]: done

File: src/index/hgraph_index.cpp (hunk 23 above)
Comment (LHT129) [reply src/index/hgraph_index.cpp]: done

File: src/index/hgraph_index.cpp (hunk 20 above)
Comment (LHT129) [reply src/index/hgraph_index.cpp]: done

File: src/index/hgraph_index.h (hunk 19 above)
Comment (LHT129) [reply src/index/hgraph_index.h]: done

File: src/index/hgraph_index.h (hunk 17 above)
Comment (LHT129) [reply src/index/hgraph_index.h:46]: It's better to move these to typing or other files, the hnswlib use it too.

File: src/index/hgraph_index.cpp [hunk 24]
```diff
@@ -0,0 +1,675 @@
...
+                                             ep,
+                                             1,
+                                             nullptr);
+        ep = result.top().second;
+    }
+
+    auto params = HnswSearchParameters::FromJson(parameters);
+
+    auto ef = params.ef_search;
+    auto results = this->search_one_graph(
```
Comment (jiaweizone) [inline src/index/hgraph_index.cpp]: results -> search_results

File: src/index/hgraph_index.cpp [hunk 25]
```diff
@@ -0,0 +1,675 @@
...
+        query->GetFloat32Vectors(), this->bottom_graph_, this->basic_flatten_codes_, ep, ef, &ft);
+
+    while (results.size() > k) {
+        results.pop();
+    }
+
+    auto result = Dataset::Make();
```
Comment (jiaweizone) [inline src/index/hgraph_index.cpp]: result -> dataset_result ?

File: src/index/hgraph_index.cpp [hunk 26]
```diff
@@ -0,0 +1,675 @@
...
+            std::unique_lock<std::mutex> lock(this->global_mutex_);
+            bool need_lock = false;
+            if (level >= int64_t(this->max_level_) || bottom_graph_->TotalCount() == 0) {
+                for (int64_t j = max_level_; j <= level; ++j) {
+                    this->route_graphs_.emplace_back(this->generate_one_route_graph());
+                }
+                max_level_ = level + 1;
+                need_lock = true;
+            } else {
+                lock.unlock();
```
Comment (jiaweizone) [inline src/index/hgraph_index.cpp:233]: Don't need a lock most of the time?

File: src/index/hnsw.cpp [hunk 27]
```diff
@@ -129,6 +129,7 @@ HNSW::build(const DatasetPtr& base) {
         std::vector<int64_t> failed_ids;
         {
             SlowTaskTimer t("hnsw graph");
+            //#pragma omp parallel for
```
Comment (jiaweizone) [inline src/index/hnsw.cpp]: add TODO ?

File: src/index/hgraph_index.cpp (hunk 24 above)
Comment (LHT129) [reply src/index/hgraph_index.cpp]: done

File: src/index/hgraph_index.cpp (hunk 25 above)
Comment (LHT129) [reply src/index/hgraph_index.cpp]: done

File: src/index/hgraph_index.cpp (hunk 26 above)
Comment (LHT129) [reply src/index/hgraph_index.cpp:233]: yes, the lock only for some global members not for detail graph nodes and neighbors

File: src/index/hnsw.cpp (hunk 27 above)
Comment (LHT129) [reply src/index/hnsw.cpp]: delete it, we use threadpool for parallel

File: src/index/hgraph_index.cpp (hunk 26 above)
Comment (jiaweizone) [reply src/index/hgraph_index.cpp:233]: OK, we can change to check and lock mode to reduce lock cost ？

**Review by jiaweizone — APPROVED** [review]
LGTM

**Review by wxyucs — APPROVED** [review]
lgtm

**Pull request merged**